File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/inheritance/super_and_direct_dispatch/super_and_direct_dispatch.kt
// KIND: STANDALONE
// FREE_COMPILER_ARGS: -opt-in=kotlin.native.internal.InternalForKotlinNative
// MODULE: Inheritance
// FILE: super_and_direct_dispatch.kt

// Non-virtual ("direct dispatch") forward bridges: a Swift subclass of an open Kotlin class must be
// able to call `super.method()` and to inherit non-overridden open methods without infinitely
// recursing through the vtable slot patched with the Swift reverse trampoline.

open class Vehicle {
    open fun describe(): String = "kotlin-vehicle"
    open fun wheels(): Int = 4
}

fun callDescribe(v: Vehicle): String = v.describe()
fun callWheels(v: Vehicle): Int = v.wheels()

File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/inheritance/super_and_direct_dispatch/super_and_direct_dispatch.swift
import Inheritance
import Testing

@Test
func swiftOverrideCanCallSuperOnKotlinClass() throws {
    // A Swift override that calls `super.method()` must reach the inherited Kotlin implementation
    // via the non-virtual ("direct dispatch") forward bridge, instead of re-entering the patched
    // vtable slot and recursing forever.
    class FancyVehicle: Vehicle {
        override func describe() -> String {
            return "fancy-" + super.describe()
        }
    }
    let v = FancyVehicle()

    // Direct Swift dispatch: the override runs and its super-call lands in Kotlin.
    #expect(v.describe() == "fancy-kotlin-vehicle")
    // Kotlin-side dispatch reaches the Swift override, whose super-call again lands in Kotlin.
    #expect(callDescribe(v: v) == "fancy-kotlin-vehicle")

    // Original Kotlin instance is unaffected.
    #expect(callDescribe(v: Vehicle()) == "kotlin-vehicle")
    #expect(Vehicle().describe() == "kotlin-vehicle")
}

@Test
func swiftSubclassInheritsNonOverriddenKotlinMethod() throws {
    // A Swift subclass that overrides only some methods must still be able to invoke the
    // non-overridden ones (whose vtable slots are also patched) without infinite recursion.
    class FancyVehicle: Vehicle {
        override func describe() -> String { "fancy" }
        // `wheels()` is intentionally not overridden.
    }
    let v = FancyVehicle()

    #expect(v.describe() == "fancy")
    // Inherited, non-overridden method via direct Swift dispatch.
    #expect(v.wheels() == 4)
    // Inherited, non-overridden method reached through a Kotlin caller must not recurse.
    #expect(callWheels(v: v) == 4)
    #expect(callDescribe(v: v) == "fancy")
}
